Index: lecture#8/tests.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.BaseRevisionTextPatchEP
<+>\r\ndef dec(if_, reason=''):\r\n    def decorator(func):\r\n        def wrapper(*args, **kwargs):\r\n            if if_:\r\n                return f'Skipping test \"{func.__name__}\". Reason: {reason}'\r\n            else:\r\n                return func(*args, **kwargs)\r\n        return wrapper\r\n    return decorator\r\n\r\n\r\n@dec(sum([10, 15]) == 25, 'becouse sum entered values is equal to 25 and so condition is met.')\r\ndef test_func(*args, **kwargs):\r\n    return sum(args, **kwargs) == 25\r\n\r\n\r\nprint(test_func())\r\n
===================================================================
diff --git a/lecture#8/tests.py b/lecture#8/tests.py
--- a/lecture#8/tests.py	
+++ b/lecture#8/tests.py	
@@ -10,7 +10,7 @@
     return decorator
 
 
-@dec(sum([10, 15]) == 25, 'becouse sum entered values is equal to 25 and so condition is met.')
+@dec(sum([10, 10, 5]) == 25, 'becouse sum entered values is equal to 25 and so condition is met.')
 def test_func(*args, **kwargs):
     return sum(args, **kwargs) == 25
 
